doc/developer-notes.md

```diff
@@ -839,6 +840,36 @@ namespace {
 #endif // BITCOIN_FOO_BAR_H
 ```
 
+Construct On First Use
+------------------------
+
+It is not possible to control the order in which non-local static objects in *different* translation units are initialized.
+This is the so-called "static initialization order problem". It could cause the use of an object:
+- before it is constructed (on program startup)
+- after it is destructed (at program exit)
+
+Both cases lead to undefined behavior.
+
+To prevent the static initialization order problem, be sure to Construct On First Use:
+
+```C++
+BCLog::Logger& LogInstance()
+{
+    // See doc/developer-notes.md#construct-on-first-use
+    static BCLog::Logger* g_logger{new BCLog::Logger()};
```

Comment:
It's nicer to use std::unique_pointer here, so that the object also gets destroyed on shutdown (in reserve order compared to initialization).

Reply:
Agree that the docs should recommend the version that doesn't leak. An additional note can say that a version that leaks might be preferable in rare circumstances.

Reply:
IMO, destroying an object via `std::unique_pointer`:
- is not required as no actual memory leak happens at program exit ([C++ FAQ](https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use)); in another words, not destroying an object does not implies memory leaks in that case
- is not safe as this object could (potentially) be destroyed before its last use at program exit sequence, e.g., instances of `BCLog::Logger` and `LockData` classes.

Reply:
The initial intention of changing the docs was to drop every long explanation in the source, and just re-direct a code reader to the docs. If these changes seem controversial, all doc-change commits could be dropped :)

Reply:
* Some tools check for leaked memory at exit, so making that the default increases verbosity for those tools.
* The destructor must be trivial to be able to do that in the first place. I am not sure why you remove that requirement in the docs.

Reply:
Not doing that is not a leak. It's just memory that's still reachable at exit (which at least valgrind classifies differently than leaked memory, and doesn't warn about).

I think it's just cleaner to clean things up too; for example this makes it easier to move globals into a class if that ever makes sense. Also, if that works, it means there exists a clear layering between which modules depend on which others in terms of initialization order, which is a low-priority nice-to-have if possible.

Reply:
Right, `unique_ptr` won't work here. It was the same destruction order (thougn not initialization order) problem as using a global directly.

Reply:
Changes to the `developer-notes.md` [dropped](https://github.com/bitcoin/bitcoin/pull/18881#issuecomment-625296903).